server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java

```
@@ -0,0 +1,176 @@
+        }
+
+        private void addWriter(EdgeTreeWriter writer) {
+            minLon = Math.min(minLon, writer.minX);
+            minLat = Math.min(minLat, writer.minY);
+            maxLon = Math.max(maxLon, writer.maxX);
+            maxLat = Math.max(maxLat, writer.maxY);
+            shapeWriters.add(writer);
+        }
+
+        @Override
+        public Void visit(GeometryCollection<?> collection) {
+            for (Geometry geometry : collection) {
+                geometry.visit(this);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visit(Line line) {
+            // TODO
+            return null;
+        }
+
+        @Override
+        public Void visit(MultiLine multiLine) {
+            for (Line line : multiLine) {
+                visit(line);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visit(Polygon polygon) {
+            // TODO (support holes)
+            LinearRing outerShell = polygon.getPolygon();
+            addWriter(new EdgeTreeWriter(asIntArray(outerShell.getLons()), asIntArray(outerShell.getLats())));
+            return null;
+        }
+
+        @Override
+        public Void visit(MultiPolygon multiPolygon) {
+            for (Polygon polygon : multiPolygon) {
+                visit(polygon);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visit(Rectangle r) {
+            int[] lats = new int[] { (int) r.getMinLat(), (int) r.getMinLat(), (int) r.getMaxLat(), (int) r.getMaxLat(),
+                (int) r.getMinLat()};
+            int[] lons = new int[] { (int) r.getMinLon(), (int) r.getMaxLon(), (int) r.getMaxLon(), (int) r.getMinLon(),
+                (int) r.getMinLon()};
+            addWriter(new EdgeTreeWriter(lons, lats));
+            return null;
+        }
+
+        @Override
+        public Void visit(Point point) {
+            // TODO
+            return null;
```

I think this should throw `UnsupportedOperationException` for now to make sure that if we ever hit this we know about it. The same for other visit methods.